### PR TITLE
cpu/atmega_common: __exit_isr thread_yield

### PR DESCRIPTION
**Pull request by Josar**
As proposed by @gebart in #8979 this PR adds `thread_yield();` to `__exit_isr` similar to #6171.

This avoids duplicated code. 
Additionaly code was uncrustified.

### CODE REVIEW

File: cpu/atmega_common/periph/uart.c [hunk 1]
```diff
@@ -166,45 +166,35 @@ void uart_poweroff(uart_t uart)
...
-    if (sched_context_switch_request) {
-        thread_yield();
-    }
+    __exit_isr();
 }
 
 #ifdef UART_0_ISR
 ISR(UART_0_ISR, ISR_BLOCK)
 {
-    __enter_isr();
```
Comment (gebart) [inline cpu/atmega_common/periph/uart.c]: I prefer the original design where enter isr is the first line of each ISR

Comment (Josar) [reply cpu/atmega_common/periph/uart.c]: Is this just a preference or are there any reasons. As i see there is no benefit in settings the `__in_isr` flag as early as possible, as there is no chance that code could be invoked which could use this information. 
Setting the flag prevents code in the callback to sleep or do other things which schould not be done in interrupts, so setting it before the callback should be sufficient, imho.

I change it as it is like this in all other interrupt handler and thought it would also increase consistency.

Comment (gebart) [reply cpu/atmega_common/periph/uart.c]: Personal preference. The consistency helps when reviewing the code and also reduces the risk of missing it if someone uses this periph driver as a template for some other driver for the same CPU.

File: cpu/atmega_common/include/cpu.h [hunk 2]
```diff
@@ -75,20 +81,20 @@ void cpu_init(void);
 /**
  * @brief   Print the last instruction's address
  */
-__attribute__((always_inline)) static inline void cpu_print_last_instruction(void)
+ static inline void __attribute__((always_inline)) cpu_print_last_instruction(void)
```
Comment (kaspar030) [inline cpu/atmega_common/include/cpu.h]: unrelated

File: cpu/atmega_common/include/cpu.h [hunk 3]
```diff
@@ -75,20 +81,20 @@ void cpu_init(void);
...
 {
     uint8_t hi;
     uint8_t lo;
     uint16_t ptr;
 
-    __asm__ volatile( "in __tmp_reg__, __SP_H__  \n\t"
+    __asm__ volatile ("in __tmp_reg__, __SP_H__  \n\t"
```
Comment (kaspar030) [inline cpu/atmega_common/include/cpu.h]: unrelated

File: cpu/atmega_common/periph/uart.c (hunk 1 above)
Comment (kaspar030) [reply cpu/atmega_common/periph/uart.c]: +1

Comment (Josar) [reply cpu/atmega_common/periph/uart.c]: @gebart maybe we just have a language problem as consistency translates into two different meanings in german. Do you mean consistency in regards as it stays at it was or consistency as it is similar to other code of the atmegas?
https://github.com/RIOT-OS/RIOT/blob/22aa6b9cc5c0ae95663209a3bdeb13e6b5630e63/cpu/atmega_common/periph/gpio.c#L264-L274
https://github.com/RIOT-OS/RIOT/blob/22aa6b9cc5c0ae95663209a3bdeb13e6b5630e63/cpu/atmega_common/periph/timer.c#L169-L189

Comment (gebart) [reply cpu/atmega_common/periph/uart.c]: I guess you are right regarding consistency. I am not very engaged in the atmega development, sorry. 
What I was trying to push for is to have a strict template for the ISR definitions which should be the same across all drivers for the CPU family, which you showed that this is. Consistency makes refactoring the driver code easier in the future. I am happy as long as this design matches the other periph drivers for the atmega.

File: cpu/atmega_common/include/cpu.h [hunk 4]
```diff
@@ -34,8 +34,10 @@
 #include <stdint.h>
 
 #include <avr/interrupt.h>
+#include "board.h"
```
Comment (gebart) [inline cpu/atmega_common/include/cpu.h]: is this include actually used?

**Comment by gebart** [comment]
My concerns have been addressed.

**Comment by Josar** [comment]
@kaspar030 can you approve it like this?

**Comment by Josar** [comment]
@kaspar030 ping 😏

File: cpu/atmega_common/include/cpu.h [hunk 5]
```diff
@@ -52,19 +53,23 @@ extern "C" {
...
 {
     __in_isr = 1;
 }
 
 /**
- * @brief Flag exiting of an ISR
+ * @brief Run this code on exiting interrupt routines
  */
-static inline void __exit_isr(void)
+static inline void __attribute__((always_inline)) __exit_isr(void)
```
Comment (kaspar030) [inline cpu/atmega_common/include/cpu.h]: why always inline?

File: cpu/atmega_common/include/cpu.h [hunk 6]
```diff
@@ -52,19 +53,23 @@ extern "C" {
 extern volatile uint8_t __in_isr;
 
 /**
- * @brief Flag entering of an ISR
+ * @brief Run this code on entering interrupt routines
  */
-static inline void __enter_isr(void)
+static inline void __attribute__((always_inline)) __enter_isr(void)
```
Comment (kaspar030) [inline cpu/atmega_common/include/cpu.h]: why always inline? Don't trust the compiler?

File: cpu/atmega_common/include/cpu.h (hunk 5 above)
Comment (Josar) [reply cpu/atmega_common/include/cpu.h]: As it will be executed in an ISR, i thought speed could matter.
And if optimization is off it won't 
> GCC does not inline any functions when not optimizing unless you specify the ‘always_inline’ attribute for the function, like this:
> 
> ```
>/* Prototype.  */
> inline void foo (const char) __attribute__((always_inline));
>```

But as the statement is the attribute has to be after the function, i am not sure but, i think it does not matter.

This [benchmark ](https://indico.cern.ch/event/386232/sessions/159923/attachments/771039/1057534/always_inline_performance.pdf) makes me think maybe `inline` only is better, nevertheless i think i forgot to remove it.

**Comment by Josar** [comment]
@gebart and @kaspar030 rebased onto #8904 and added neccessary changes.

**Comment by ZetaR60** [comment]
This looks okay to me, but note that it introduces a yield in the periph/gpio.c ISR where there previously was none.

**Comment by Josar** [comment]
@ZetaR60 i think that is not fully correct, it request a thread yield if there was a `sched_context_switch_request` in the callback handler. Which i would expect to happen, as the `__in_isr` blocks the context switch and at the end of an isr it has to be executed. Otherwise it would be lost or executed with an other thread_yield.

But thinking further is it neccessary to duplicate the code?

```
static inline void __exit_isr(void)
{
     if (sched_context_switch_request) {
         thread_yield();
        thread_yield_isr();
     }
   __in_isr = 0;
}
```
`thread_yield_isr();` is the same as `thread_yield_higher();` so just bringing `__in_isr = 0` before the `if` and setting reti to the end will have the same result with less code, or not ?

```
static inline void __exit_isr(void)
{
     __in_isr = 0;
     if (sched_context_switch_request) {
         thread_yield();
     }
    __asm__ volatile ("reti");
}
```

**Comment by ZetaR60** [comment]
Yes, you are right about my comment being erroneous. Sorry for the lazy thought process.

There are problems with moving ```__asm__ volatile("reti");``` to a different area from ```__context_restore();``` because the context change directly manipulates the program counter, and returning from a function also manipulates the program counter. You would have to add ```__attribute__((always_inline)) inline``` to thread_yield / thread_yield_isr, which means you would have a copy of all the contents of __context_save and __context_restore in every ISR.

**Comment by Josar** [comment]
@ZetaR60 seems it was too late for my brain, of course moving `reti` does not work.

Maybe we could agree on just putting the content of the `if` to thread_yield_isr();` ?

I don`t imagine a situation where saving the context and running the scheduler is necessary if there was not a thread change request. 

Further thoughts, I think there should be a function like `sched_thread_change_needed` to avoid saving and restoring the context when a yield or yield higher is issued but there is no higher.
As of now `sched_run` checks this, after the context was allready saved.

**Comment by ZetaR60** [comment]
@Josar can you be a bit more precise which things you are referring to? Currently nothing is run at the end of an ISR if there was no attempt to yield within the callback, because sched_context_switch_request is not set. Are the other things you are mentioning proposed changes to core/thread.c ? Currently this PR is a refactor; perhaps discussion of changes of behavior should be moved to another PR specific to that change? There might be some improvements to be made in the way I rewrote context swapping, but that rewrite took a lot of testing to do properly.

**Comment by ZetaR60** [comment]
Those sound like reasonable changes. They would involve changing behavior though, so the extensive testing done in #8904 probably ought to be re-checked in a PR that changes it.

**Comment by Josar** [comment]
@ZetaR60 thinking further i think this can be optimized.

After deciding that a context swap is allowed in an ISR there is no need of preventing an context swap in an ISR, the only thing which has to be taken care of is to return with `reti`. The rest are remains which only add overhead.

As it is now we check for isr set the flag, return to caller, call next function, check for flag and do the swap.
Removing this will not only encrease swapping in interrupts but also increase interrupts with no yield as the flag check is removed. 

```
void thread_yield_higher(void) {
        __context_save();
        sched_run();
        __context_restore();

    if (irq_is_in() == 0) {
        __asm__ volatile("ret");
    } else {
        __exit_isr();
        __asm__ volatile ("reti");
    }
}
```

and removing the if irq in the scheduler

```
    if (!on_runqueue || (current_prio > other_prio)) {
        // if (irq_is_in()) {
        //     DEBUG("sched_switch: setting sched_context_switch_request.\n");
        //     sched_context_switch_request = 1;
        // }
        // else {
            DEBUG("sched_switch: yielding immediately.\n");
            thread_yield_higher();
        // }
    }
    else {
        DEBUG("sched_switch: continuing without yield.\n");
    }
```

first test with `test/xtimer_msg` shows expected results.

**Comment by Josar** [comment]
@kYc0o would you mind having a look at this?

**Comment by ZetaR60** [comment]
@Josar This change to thread_yield_higher is a pretty major change in behavior. Could you create another PR for it?

**Comment by Josar** [comment]
@kaspar030 any further remarks?

**Comment by Josar** [comment]
@kaspar030 some further requests?

**Comment by Josar** [comment]
@kaspar030 maybe you could find some time?

**Comment by ZetaR60** [comment]
I ran some tests, and got a couple failures:

```
thread_basic			OK
thread_cooperation		OK
thread_exit			OK
thread_flags			OK
thread_flags_xtimer		Reports failure
thread_flood			OK
thread_msg			OK
thread_msg_block_wo_queue	OK
thread_msg_block_w_queue	OK
thread_msg_seq			OK
xtimer_hang			OK
xtimer_periodic_wakeup		OK
xtimer_remove			Reports failure
xtimer_reset			OK
xtimer_usleep			OK
xtimer_usleep_short		OK
```

I think we need a bit more extensive examination. @Josar I suggest rebasing on trunk and running tests side by side with your code and the exact same revision without your code.

**Comment by Josar** [comment]
Rebased and fixed.
All above Tests run successfull.

PS: it would be nice to have a readme in each  test with the expected output.

**Comment by ZetaR60** [comment]
Confirmed that the previously failed tests are now succeeding on mega-xplained. Also did another proofread of the changes, and they look fine. Let me know if I should run any other tests on mega-xplained.

**Comment by Josar** [comment]
@kaspar030 any further remarks?

**Comment by Josar** [comment]
@kaspar030 could you have a second look?

**Comment by ZetaR60** [comment]
@Josar You are being a bit aggressive with pinging Kaspar. I know things often move slowly, but I don't think repeating requests that much is going to help.

**Comment by kYc0o** [comment]
@Josar please squash.

**Comment by kYc0o** [comment]
All the tests in https://github.com/RIOT-OS/RIOT/pull/8988#issuecomment-392869767 passed on arduino-mega2560.

**Comment by Josar** [comment]
I uncrustified the touched files, hope this is allright.

File: cpu/atmega_common/periph/timer.c [hunk 7]
```diff
@@ -87,17 +88,21 @@ int timer_init(tim_t tim, unsigned long freq, timer_cb_t cb, void *arg)
     uint8_t pre = 0;
 
     /* make sure given device is valid */
-    if (tim >= TIMER_NUMOF) {
+    if (tim >= TIMER_NUMOF)
+    {
```
Comment (aabadie) [inline cpu/atmega_common/periph/timer.c]: Is this change the result of uncrustify ? Then I think there's an issue with the uncrustify configuration file. The opening curly brace should be on the same line as the if.

**Comment by Josar** [comment]
@aabadie i think that the uncrustifying went right this time.
@kYc0o rebased and squashed.

**Comment by kYc0o** [comment]
Perfect! Thanks for your contribution!
